gazebo/profiler/Profiler.hh

```diff
@@ -0,0 +1,165 @@
+    /// adding inspection points easier.
+    ///
+    /// * IGN_PROFILE_THREAD_NAME - Set the name of the current profiled thread.
+    /// * IGN_PROFILE_LOG_TEXT - Log text to the profiler console (if supported)
+    /// * IGN_PROFILE_BEGIN - Begin a named profile sample
+    /// * IGN_PROFILE_END - End a named profile sample
+    /// * IGN_PROFILE - RAII-style profile sample. The sample will end at the
+    ///     end of the current scope.
+    class GZ_PROFILER_VISIBLE Profiler
+        : public virtual SingletonT<Profiler>
+    {
+      /// \brief Constructor
+      protected: Profiler();
+
+      /// \brief Destructor
+      protected: ~Profiler();
+
+      /// \brief Set the name of the current thread
+      /// \param[in] _name Name to set
+      public: void SetThreadName(const char *_name);
+
+      /// \brief Log text to profiler output (if supported)
+      /// If the underlying profiler implementation supports additional
+      /// log messages, this can be used to send.
+      ///
+      /// Currently, the Remotery implentation supports this functionality.
+      /// \param[in] _text Text to log.
+      public: void LogText(const char *_text);
+
+      /// \brief Begin a named profiling sample.
+      /// Begins a CPU profiler sample with a given name. Can optionally take
+      /// a hash parameter to be cached between executions of `BeginSample`, so
+      /// that hashes don't need to be recomputed.
+      /// \param[in] _name Name of the sample
+      /// \param[in,out] _hash An optional hash value that can be cached
+      ///   between executions.
+      public: void BeginSample(const char *_name, uint32_t *_hash = nullptr);
+
+      /// \brief End a profiling sample.
+      public: void EndSample();
+
+      /// \brief Get the underlying profiler implentation name
+      public: std::string ImplementationName() const;
+
+      /// \brief Detect if profiler is enabled and has an implementation
+      public: bool Valid() const;
+
+      /// \brief Pointer to the profiler implementation
+      private: ProfilerImpl *impl;
+
+      /// \brief Needed for SingletonT.
+      private: friend class SingletonT<Profiler>;
+    };
+
+    /// \brief Used to provide C++ RAII-style profiling sample.
+    /// The sample will start on the construction of the `ScopedProfile` object
+    /// and stop when the object leaves scope.
+    class GZ_PROFILER_VISIBLE ScopedProfile
+    {
+      /// \brief Constructor. Starts profile sample.
+      /// \param[in] _name Name of the sample
+      /// \param[in,out] _hash An optional hash value that can be cached
+      ///   between executions.
+      public: ScopedProfile(const char *_name, uint32_t *_hash)
+      {
+        Profiler::Instance()->BeginSample(_name, _hash);
+      }
+
+      /// \brief Destructor. Stops profile sample.
+      public: ~ScopedProfile()
+      {
+        Profiler::Instance()->EndSample();
+      }
+    };
+  }
+}
+
+#ifndef IGN_PROFILER_ENABLE
```

Comment:
Can we change all of these from `IGN` to `GZ`? Just to be safe in case the 2 of them ever get mixed in the same project.